src/main/java/io/cdap/plugin/couchbase/source/JsonObjectToRecordTransformer.java

```diff
@@ -0,0 +1,240 @@
+      JsonObject payload = jsonObject.getObject(config.getBucket());
+      if (jsonObject.size() == 1) {
+        return payload;
+      }
+      // Result object can contain other fields for queries such as: SELECT meta(`test-bucket`).id, * from `test-bucket`
+      // Include them as well
+      jsonObject.removeKey(config.getBucket());
+      for (String name : jsonObject.getNames()) {
+        payload.put(name, jsonObject.get(name));
+      }
+      return payload;
+    }
+
+    return jsonObject;
+  }
+
+  private StructuredRecord extractRecord(@Nullable String fieldName, JsonObject object, Schema schema) {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    List<Schema.Field> fields = Objects.requireNonNull(schema.getFields(), "Schema fields cannot be empty");
+    for (Schema.Field field : fields) {
+      // Use full field name for nested records to construct meaningful errors messages.
+      // Full field names will be in the following format: 'record_field_name.nested_record_field_name'
+      String fullFieldName = fieldName != null ? String.format("%s.%s", fieldName, field.getName()) : field.getName();
+      Schema nonNullableSchema = field.getSchema().isNullable() ? field.getSchema().getNonNullable()
+        : field.getSchema();
+      Object value = extractValue(fullFieldName, object.get(field.getName()), nonNullableSchema);
+      builder.set(field.getName(), value);
+    }
+    return builder.build();
+  }
+
+  private Object extractValue(String fieldName, Object object, Schema schema) {
+    if (object == null) {
+      return null;
+    }
+
+    Schema.LogicalType fieldLogicalType = schema.getLogicalType();
+    if (fieldLogicalType != null) {
+      switch (fieldLogicalType) {
+        case DECIMAL:
+          ensureTypeValid(fieldName, object, Number.class);
+          return extractDecimal(fieldName, (Number) object, schema);
+        default:
+          throw new UnexpectedFormatException(String.format("Field '%s' is of unsupported type '%s'", fieldName,
+                                                            fieldLogicalType.name().toLowerCase()));
+      }
+    }
+
+    Schema.Type fieldType = schema.getType();
+    switch (fieldType) {
+      case BOOLEAN:
+        ensureTypeValid(fieldName, object, Boolean.class);
+        return object;
+      case INT:
+        ensureTypeValid(fieldName, object, Integer.class);
+        return object;
+      case DOUBLE:
+        ensureTypeValid(fieldName, object, Double.class);
+        return object;
+      case LONG:
+        ensureTypeValid(fieldName, object, Integer.class, Long.class);
+        return ((Number) object).longValue();
+      case STRING:
+        ensureTypeValid(fieldName, object, String.class);
+        return object;
+      case MAP:
+        ensureTypeValid(fieldName, object, JsonObject.class);
+        Schema valueSchema = schema.getMapSchema().getValue().isNullable()
```

Comment:
nit: more readable to break this up into 2 statements instead of repeating schema.getMapSchema().getValue()

Reply:
Done.